seismicpro/src/metrics/qc_metric.py

```diff
@@ -0,0 +1,584 @@
+                                          gather.samples).astype(int)
+        s_beg, s_end = max(i1, i2), min(i3, i4)
+
+        n_begs = np.full(gather.n_traces, fill_value=n_beg, dtype=int)
+        s_begs = np.full(gather.n_traces, fill_value=s_beg, dtype=int)
+        win_size = min(n_end - n_beg, s_end - s_beg)
+
+        return n_begs, s_begs, win_size
+
+    @staticmethod
+    def _get_res(gather, offsets, win_size=100, n_start=10, s_start=1500, **kwargs):
+        """QC indicator implementation."""
+
+        mask = ((gather.offsets >= offsets[0]) & (gather.offsets <= offsets[1]))
+
+        n_begs, s_begs, win_size = TraceSinalToNoiseRMSRatio._get_indices(gather, win_size, n_start, s_start, mask)
+
+        s_begs[~mask] = -1
+        n_begs[~mask] = -1
+
+        res = rms_2_windows_ratio(gather.data, n_begs, s_begs, win_size)
+
+        return res
+
+    def plot(self, coords, ax, offsets, win_size=100, n_start=10, s_start=1500, **kwargs):
+        """Gather plot sorted by offset with tracewise indicator on a separate axis and signal and noise windows"""
+        gather = self.survey.get_gather(coords)
+
+        res = self.get_res(gather, from_headers=False, offsets=offsets,
+                           win_size=win_size, n_start=n_start, s_start=s_start, **kwargs)
+        res = self.aggr(res, tracewise=True)
+
+        gather = self.preprocess(gather, **kwargs)
+        order = np.argsort(gather.offsets.ravel(), kind='stable')
+
+        gather.sort(by='offset').plot(ax=ax)
+
+        divider = make_axes_locatable(ax)
+        top_ax = divider.append_axes("top", sharex=ax, size="12%", pad=0.05)
+        top_ax.xaxis.set_visible(False)
+        top_ax.set_yscale(self.top_ax_y_scale)
+
+        top_ax.plot(res[order], '.--')
+        if self.threshold is not None:
+            top_ax.axhline(self.threshold, alpha=0.5)
+
+        mask = (gather.offsets >= offsets[0]) & (gather.offsets <= offsets[1])
+        offs_ind = np.nonzero(mask)[0]
+
+        n_begs, s_begs, win_size = TraceSinalToNoiseRMSRatio._get_indices(gather, win_size, n_start, s_start, mask)
+
+        n_rec = (offs_ind[0], n_begs[0]), len(offs_ind), win_size
+        ax.add_patch(patches.Rectangle(*n_rec, linewidth=1, edgecolor='magenta', facecolor='none'))
+        s_rec = (offs_ind[0], s_begs[0]), len(offs_ind), win_size
+        ax.add_patch(patches.Rectangle(*s_rec, linewidth=1, edgecolor='lime',facecolor='none'))
+
+        set_title(top_ax, gather)
+
+class TraceSinalToNoiseRMSRatioAdaptive(TracewiseMetric):
+    """ Signal to Noise RMS ratio computed using provided windows """
+    name = "trace_RMS_Ratio_Adaptive"
+    is_lower_better = False
+    threshold = None
+    top_ax_y_scale = 'log'
+    views = 'plot'
+
+    @staticmethod
+    def _get_indices(gather,  win_size, shift_up, shift_down):
+        """Convert times to use for noise and signal windows into indices"""
+        fbp = gather.headers[HDR_FIRST_BREAK].values
```

Comment:
same as before. What if fbp stores in diff header and what if fbp are missing?
```suggestion
        fbp = gather[HDR_FIRST_BREAK].ravel()
```

Reply:
added `first_breaks_col` parameter and explicit error message if the column is missing in headers